File: old/Probleme183.py
#~/usr/bin/python

import string
import sys
import math
import time
import decimal

def pgcd(a,b):
	if b == 0:
		return a
	else:
		return pgcd(b,a%b)

def power(a,b):
	if b == 0:
		return 1
	elif b == 1:
		return a
	elif b%2 == 0:
		return power(a**2,b/2)
	else:
		return a * power(a,b-1)

def P(n,k):
	return decimal.Decimal(str(n/k))**int(k)

def infini(n,k):
	d = pgcd(n,k)
	n1 = n / d
	k1 = k / d
	p = 2
	while k1 > 1:
		if k1%p == 0:
			if p not in [2,5]:
				return True
			else :
				k1 = k1 / p
		elif p > 5:
			return True
		else:
			p = p + 1
	return False

def terminating(n):
	k0 = decimal.Decimal(str(n))/decimal.Decimal(str(math.e))
	k_max = math.ceil(k0)
	p_max = P(n,k_max)
	k_min = math.floor(k0)
	p_min = P(n,k_min)
	k = 0
	if (p_max > p_min):
		k = k_max
	else:
		k = k_min
	return not infini(n,k)

def problem183():
	limite = int(sys.argv[1])
	sigma = 0
	print 11, terminating(11)
	print 8, terminating(8)
	for i in xrange(5,limite + 1):
		#print i
		if terminating(i):
			sigma = sigma - i
		else:
			sigma = sigma + i
	return sigma

print problem183()
	
	
	
	

File: probleme1xx/probleme183.cpp
#include "problemes.h"
#include "arithmetiques.h"
#include "utilitaires.h"

#include <iostream>
#include <fstream>
#include <algorithm>
#include <limits>

#include <boost/rational.hpp>

typedef unsigned long long nombre;
typedef std::vector<nombre> vecteur;

typedef boost::rational<nombre> fraction;

namespace
{
    bool infini(nombre n, nombre k)
    {
        k /= arithmetiques::PGCD(n, k);
        while (k%2 == 0)
            k /= 2;
        
        while (k%5 == 0)
            k /= 5;
            
        return k == 1;
    }
    
    bool terminating(nombre n)
    {
        long double k0 = n / M_E;
        nombre k_max = static_cast<nombre>(std::ceil(k0));
        nombre k_min = static_cast<nombre>(std::floor(k0));

        long double p_max = puissance::puissance(n / std::ceil(k0), k_max);
        long double p_min = puissance::puissance(n / std::floor(k0), k_min);
        
        if (p_max > p_min)
            return infini(n, k_max);
        else
            return infini(n, k_min);
    }
}

ENREGISTRER_PROBLEME(183, "Maximum product of parts")
{
    // Let N be a positive integer and let N be split into k equal parts, r = N/k, so that N = r + r + ... + r.
    // Let P be the product of these parts, P = r × r × ... × r = rk.
    //
    // For example, if 11 is split into five equal parts, 11 = 2.2 + 2.2 + 2.2 + 2.2 + 2.2, 
    // then P = 2.25 = 51.53632.
    //
    // Let M(N) = Pmax for a given value of N.
    //
    // It turns out that the maximum for N = 11 is found by splitting eleven into four equal parts
    // which leads to Pmax = (11/4)4; that is, M(11) = 14641/256 = 57.19140625, which is a terminating decimal.
    //
    // However, for N = 8 the maximum is achieved by splitting it into three equal parts, so M(8) = 512/27,
    // which is a non-terminating decimal.
    //
    // Let D(N) = N if M(N) is a non-terminating decimal and D(N) = -N if M(N) is a terminating decimal.
    //
    // For example, ΣD(N) for 5 ≤ N ≤ 100 is 2438.
    //
    // Find ΣD(N) for 5 ≤ N ≤ 10000.
    nombre limite = 10000;
    // std::cout << std::boolalpha;
    // std::cout << "terminating(8) = " << terminating(8) << std::endl;
    // std::cout << "terminating(11) = " << terminating(11) << std::endl;
    
    nombre resultat_positif = 0;
    nombre resultat_negatif = 0;
    for (nombre n = 5; n < limite + 1; ++n)
    {
        if (terminating(n))
            resultat_negatif += n;
        else
            resultat_positif += n;
    }

    return std::to_string(resultat_positif - resultat_negatif);
}
